core/utils/xml.js

```diff
@@ -76,5 +84,29 @@ Blockly.utils.xml.textToDomDocument = function(text) {
  */
 Blockly.utils.xml.domToText = function(dom) {
   var oSerializer = new XMLSerializer();
-  return oSerializer.serializeToString(dom);
+  return Blockly.utils.xml.sanitizeText(oSerializer.serializeToString(dom));
+};
+
+/**
+ * Sanitizes a text string so that it can be serialized to XML. Specifically,
+ * this replaces control characters with their NCR (numerical character
+ * reference).
+ * @param {string} text The text to sanitize.
+ * @return {string} The sanitized text.
+ */
+Blockly.utils.xml.sanitizeText = function(text) {
+  if (!text.match(Blockly.utils.xml.INVALID_CONTROL_CHARS)) {
```

Comment:
"[Do you need this? B]ecause without it, the function should still return the input text when it contains no invalid control characters."

Reply:
Good point! There is one argument for it that I can think of for it. Which is that [returning early](https://medium.com/swlh/return-early-pattern-3d18a41bba8) makes it more clear this function only cares about strings w/ control characters. But it is definitely not necessary.

Reply:
`Blockly.utils.xml.INVALID_CONTROL_CHARS.test(text)` should be much faster than `match` since it only returns true or false, and doesn't have to build up a reporting object.